### PR TITLE
Account for pixel density when masking images

### LINKED ISSUE #6770: Masking behaviour change between 1.4.1 and latest
### Most appropriate sub-area of p5.js?

- [ ] Accessibility
- [ ] Color
- [X] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [ ] Image
- [ ] IO
- [ ] Math
- [ ] Typography
- [ ] Utilities
- [ ] WebGL
- [ ] Build process
- [ ] Unit testing
- [ ] Internationalization
- [ ] Friendly errors
- [ ] Other (specify if possible)

### p5.js version

1.9.0

### Web browser and version

chrome

### Operating system

irrelevant/webEditor

### Steps:
1. old version working with 1.4.1: https://jsfiddle.net/8b253Ly1/
2. new version no longer working , just silently failing: https://editor.p5js.org/jacopom/sketches/7SjQBqBMy
3. same code, different lib version linked

### Snippet:

```js

var img;
var masked;

var mask;

function setup() {
    createCanvas(400, 400);

		// creates an invisible graphic buffer
		// that contains a rectangle
    img = createGraphics(200, 200);
    img.rect(0, 0, 100, 100);

		// creates an invisible graphic buffer
		// that contains a circle
    mask = createGraphics(200, 200);
    mask.ellipse(100, 100, 100, 100);

		// we apply the circle in 'mask' as a clipping mask
		// to the rectangle image contained in 'img'
		// we save the result in a third buffer: 'masked'
    ( masked = img.get() ).mask( mask.get() );
}

function draw() {
    background(130,0,10);

		// print the buffer after the clipping
    image(masked, 0, 0);
}

```

This is something I used to teach in class, now it's broken and doesn't even give an error, what changed? how do I fix it?

### PR DESCRIPTION
**Pull request by Papershine**
Resolves #6770

#### Changes:
Previously, image pixel densities were hard coded to be 1, but when image pixel densities were changed to be variable, the mask function was not updated, resulting in incorrectly sized images being masked. By taking into account the pixel density when calculating the actual pixel width and height for masking, the issue should be fixed.

#### Screenshots of the change:
On a high pixel density display, the previous behaviour of the code snippet in #6770 has been restored:
Screenshot 2024-02-03 at 6 29 10 PM

#### PR Checklist

- [x] `npm run lint` passes
- [x] [Unit tests] are included / updated

[Unit tests]: https://github.com/processing/p5.js/tree/main/contributor_docs#unit-tests

### CODE REVIEW

**Comment by Papershine** [comment]
Potential changes needed:

- If we try to mask two images with different pixel densities, things don't seem to work well currently (which makes sense). Potentially, we might want to upscale an image if we see that the pixel densities of the two images are different?

- May also want to add a unit test to test non-1 pixel densities - please let me know if that is necessary, thanks!

File: src/image/p5.Image.js [hunk 1]
```diff
@@ -878,21 +878,22 @@ p5.Image = class {
     }
     const currBlend = this.drawingContext.globalCompositeOperation;
 
-    let scaleFactor = 1;
+    let imgScaleFactor = this._pixelDensity;
+    let maskScaleFactor = 1;
     if (p5Image instanceof p5.Renderer) {
-      scaleFactor = p5Image._pInst._pixelDensity;
+      maskScaleFactor = p5Image._pInst._pixelDensity;
```
Comment (davepagurek) [inline src/image/p5.Image.js:884]: `p5.Image` also has a pixel density too now, maybe we can add a branch to handle when the input is an image too? https://github.com/processing/p5.js/blob/37d3324b457b177319ed65468201f2806a66eff5/src/image/p5.Image.js#L150

File: src/image/p5.Image.js [hunk 2]
```diff
@@ -878,21 +878,22 @@ p5.Image = class {
...
     }
 
     const copyArgs = [
       p5Image,
       0,
       0,
-      scaleFactor * p5Image.width,
-      scaleFactor * p5Image.height,
+      maskScaleFactor * p5Image.width,
```
Comment (davepagurek) [inline src/image/p5.Image.js:891]: Do we have to divide the factor by the the current image's scale factor to handle the case when an image (e.g. from a 2x density `p5.Graphics` after calling `.get()` on it) has density other than 1?

**Review by davepagurek — COMMENTED** [review]
Thanks for taking this on! I think those two ideas you mentioned would be great, and a test will help make sure we don't accidentally break this again in the future. We have a new visual test system (https://github.com/processing/p5.js/blob/main/contributor_docs/unit_testing.md#visual-tests), maybe that could be a good fit for a test for this?

**Comment by Papershine** [comment]
I'm trying to set up and test masking when manually changing the pixel density, but things keep bugging out. Looking into the code for `pixelDensity`, I think there are some problems here

https://github.com/processing/p5.js/blob/304ee9008d64d842ddfb4a62999a5cd62c944316/src/image/p5.Image.js#L256-L260

If the original pixel density is 2, and we want to make it 4, it seems setting `pixelDensity(4)` would break the correct width and height of the image. Is it something that should be fixed?

**Comment by davepagurek** [comment]
I think the idea is that when you import e.g. an 800x800 image from a file, we have no idea what the density is, so we assume 1x density. But if we tell p5 that actually it should be treated like a 400x400 image at 2x density, then it will change its width and height to accommodate. So in this code path, we don't want to change any of the underlying pixels, just reinterpret it.

I'll be different if you create it from a `p5.Graphics` though: for that, since it's a canvas that you draw to, we give you a new blank canvas at the size you ask for. So if you start with an 800x800 canvas and then tell it to use a pixel density of 2, under the hood, you get 1600x1600 pixels under the hood. Then if you call `.get()` on that canvas, you should also end up with an image that also has width and height of 800 at density 2, for a total of 1600x1600 pixels.

So for tests, I think you'd either (1) make a graphic at the size and density you want and then turn it into an image with `.get()`, or (2) start with an image with the full number of *pixels* you want, and then tell p5 about its density to scale down the width/height.

**Comment by taliacotton** [comment]
Hi team! Wondering if there's a plan to integrate this update into the newest release of p5. Would be grateful! Thank you!!

**Comment by davepagurek** [comment]
@Papershine are you interested in continuing to finish up the tests? No worries if not, we can always continue development from what you've started!

**Comment by Papershine** [comment]
Hey @davepagurek, sorry about the delay! I've been extremely busy with school and still doing exams right now, it would be great if anybody interested could take over and finish this PR, thank you so much! If nobody is interested, I could work on the tests in a week when my exams are done.

**Comment by Papershine** [comment]
I've updated the unit test to test for higher pixel densities (I tried to run this test on master and it does fail on master, and it passes on my branch). However, I'm not too sure how exactly we want to deal with the case where the base image and the mask have different densities. Right now I am considering scaling up/down the mask if the density is different, by first resizing the mask, and then calling `pixelDensity` to match the density of the base image. This way, we aren't inadvertently changing the pixel density of the original image. How does this sound?

**Comment by Papershine** [comment]
A test for pixel density 2, as well as a test for the case where pixel densities aren't the same, have been added.

**Comment by davepagurek** [comment]
Thanks @Papershine, the tests so far look great! A `createGraphics` test would be good too just to make sure we've handled everything, then I think we're good to go!

**Comment by Papershine** [comment]
Just added that!

**Review by davepagurek — APPROVED** [review]
Thanks for the code and the tests!

**Comment by davepagurek** [comment]
@all-contributors please add @Papershine for code, test

**Pull request merged**